File: Structural/Bridge/python/BridgePattern.py
from abc import ABC, abstractmethod

class Implementation(ABC):

    @abstractmethod
    def operation_a(self) -> str:
        pass
    
    @abstractmethod
    def operation_b(self, num: int):
        pass

class Abstraction:

    def __init__(self, implementation: Implementation):
        self.implementation = implementation
    
    def operation_one(self):
        self.implementation.operation_a()
        self.implementation.operation_b(1)
    
    def operation_two(self):
        self.implementation.operation_a()
    

class ConcreteImplementationAlpha(Implementation):

    def operation_a(self) -> str:
        print("Alpha a")
    
    def operation_b(self, num: int):
        print("Alpha b {0}".format(num))

class ConcreteImplementationBetha(Implementation):

    def operation_a(self) -> str:
        print("Betha a")
    
    def operation_b(self, num: int):
        print("Betha b {0}".format(num))

class ExtendedAbstraction(Abstraction):

    def operation_tree(self):
        self.implementation.operation_b(2)

def main():
    abstractions = [Abstraction(ConcreteImplementationAlpha()), 
    Abstraction(ConcreteImplementationBetha())]

    for abstraction in abstractions:
        abstraction.operation_one()


if __name__ == "__main__":
    main() 

File: Structural/Bridge/scala/BridgePattern.scala
trait Implementation {

    def operationA(): Unit

    def operationB(num: Int): Unit
}

class Abstraction(val implementation: Implementation) {

    def operationOne() = {
        this.implementation.operationA()
        this.implementation.operationB(1)
    }

    def operationTwo() = {
        this.implementation.operationA()
    }
}

class ConcreteImplementationAlpha extends Implementation {

    override def operationA() = println("Concrete Alpha operation A")

    override def operationB(num: Int) = println(s"Concrete Alpha operation B: $num")
}

class ConcreteImplementationBeta extends Implementation {

    override def operationA() = println("Concrete Beta operation A")

    override def operationB(num: Int) = println(s"Concrete Beta operation B: $num")
}

class ExtendedAbstraction(override val implementation: Implementation) extends Abstraction(implementation) {

    def operationThree(): Unit = this.implementation.operationB(3)
}

object BridgePattern extends App {

    val abstractionWithConcreteImplAlpha = new Abstraction(new ConcreteImplementationAlpha)
    val abstractionWithConcreteImplBeta = new Abstraction(new ConcreteImplementationBeta)
    val extendedAbstractionWithConcreteImplAlpha = new ExtendedAbstraction(new ConcreteImplementationAlpha)

    abstractionWithConcreteImplAlpha.operationOne()
    abstractionWithConcreteImplBeta.operationTwo()
    extendedAbstractionWithConcreteImplAlpha.operationThree()
}

